src/v2/providers/storage.ts

```diff
@@ -0,0 +1,419 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import { firebaseConfig } from '../../config';
+import { CloudEvent, CloudFunction } from '../core';
+import * as options from '../options';
+
+/** @hidden */
+export const provider = 'google.cloud.storage';
+/** @hidden */
+export const service = 'storage.googleapis.com';
+/** @hidden */
+export const archivedEvent = 'object.v1.archived';
+/** @hidden */
+export const finalizedEvent = 'object.v1.finalized';
+/** @hidden */
+export const deletedEvent = 'object.v1.deleted';
+/** @hidden */
+export const metadataUpdatedEvent = 'object.v1.metadataUpdated';
+
+/** StorageOptions extend EventHandlerOptions with a bucket name  */
+export interface StorageOptions extends options.EventHandlerOptions {
+  bucket?: string;
+}
+
+/** Handle a storage object archived */
+export function onObjectArchived(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object archived */
+export function onObjectArchived(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object archived */
+export function onObjectArchived(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectArchived(
+  optsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  if (arguments.length === 1) {
+    handler = optsOrHandler as (
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, archivedEvent);
+  }
+  return _onOperation(
+    handler,
+    archivedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
```

Comment:
Hmmm, it looks like we repeat this block for each of the four event type. It would be cool if we can contain all "make sure we handle `optOrHandler` appropriately in `_onOperation`; can we start a discussion in our chat room to see if there's a clever TS thing we can do?

Reply:
I've cleaned it up and moved the logic into `_onOperation`. The 4 functions now only serve to inject the correct event type.